### PR TITLE
Refactor Data Generator

### PR DESCRIPTION
**Pull request by ghost**
Changed the generating methods and added string generator.

Signed-off-by: Zongyu zy728@york.ac.uk

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/Llama-Group/Project-Llama/pull/98" target="_blank">ReviewNinja

File: llama_cpp/example/data_generator/main.cpp [hunk 1]
```diff
@@ -15,16 +15,18 @@
...
 
 int main(int argc, const char * argv[]) {
     std::vector<int> a;
-    std::vector<bool> b;
-    std::vector<double> c;
+    std::vector<double> b;
+    std::vector<std::string> c;
     llama::RandomData mRandomData = llama::RandomData();
-    int ct = 20;
+    int ct = 10;
```
Comment (0xsuu) [inline llama_cpp/example/data_generator/main.cpp]: Bad variable name. Should be more informative.

File: llama_cpp/src/Utility/DataGenerator.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+#include "Utility/DataGenerator.h"
+
+#include <vector>
+#include <random>
+#include <string>
+
+template<>
+void llama::RandomData::generateRandomData<int>(std::vector<int> *targetVector, int count) {
+    std::random_device rd_int;
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.cpp]: Bad variable name. Should be intRandomDevice or randomDeviceInt or mRandomDeviceInt.

Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.cpp]: This apply to all the variable names containing underscore.

File: llama_cpp/src/Utility/DataGenerator.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    std::uniform_real_distribution<double> ud(1, 100);
+
+    for (int i = 0; i < count; i++) {
+        targetVector->push_back(ud(re_double));
+    }
+}
+
+template<>
+void llama::RandomData::generateRandomData<std::string>(std::vector<std::string> *targetVector, int count) {
+    static const char alpha[] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.cpp]: Consider adding special characters. e.g. . _ + = - !@$@%@#^@&#**%^@

File: llama_cpp/src/Utility/DataGenerator.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    for (int i = 0; i < count; i++) {
+        targetVector->push_back(ui(re_int));
+    }
+}
+
+template<>
+void llama::RandomData::generateRandomData<double>(std::vector<double> *targetVector, int count) {
+    std::random_device rd_double;
+    std::mt19937 re_double(rd_double());
+    std::uniform_real_distribution<double> ud(1, 100);
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.cpp]: Bad variable ud. 

Can have a larger range, scaling from minimum to maximum of the 'double' type. There's a std function can do that.

File: llama_cpp/src/Utility/DataGenerator.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+#include <vector>
+#include <random>
+#include <string>
+
+template<>
+void llama::RandomData::generateRandomData<int>(std::vector<int> *targetVector, int count) {
+    std::random_device rd_int;
+    std::mt19937 re_int(rd_int());
+    std::uniform_int_distribution<int> ui(1, 100);
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.cpp]: Same as double, a larger scale would be better.

File: llama_cpp/src/Utility/DataGenerator.cpp [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        std::random_device rd_string;
+        std::mt19937 re_string(rd_string());
+        std::uniform_int_distribution<int> us(1, 10);
+
+        for (int i = 0; i < us(re_string); i++) {
+            std::random_device rd_string_index;
+            std::mt19937 re_string_index(rd_string_index());
+            std::uniform_int_distribution<int> us_index(0, 51);
+            generateStr += alpha[us_index(re_string_index)];
+            }
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.cpp]: Bracket not aligned to 'for'.

File: llama_cpp/src/Utility/DataGenerator.cpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+template<>
+void llama::RandomData::generateRandomData<std::string>(std::vector<std::string> *targetVector, int count) {
+    static const char alpha[] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+    for (int i = 0; i < count; i++) {
+        std::string generateStr;
+        std::random_device rd_string;
+        std::mt19937 re_string(rd_string());
+        std::uniform_int_distribution<int> us(1, 10);
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.cpp]: Is string length here sounds good to you? @BlueCocoa

File: llama_cpp/src/Utility/DataGenerator.cpp [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+        for (int i = 0; i < us(re_string); i++) {
+            std::random_device rd_string_index;
+            std::mt19937 re_string_index(rd_string_index());
+            std::uniform_int_distribution<int> us_index(0, 51);
+            generateStr += alpha[us_index(re_string_index)];
+            }
+
+        targetVector->push_back(generateStr);
+        }
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.cpp:98]: Bracket not aligned to 'for'.

**Comment by 0xsuu** [comment]
Please fix travis build fail, possibly the cpp file in example is not removed, do a double check.

**Comment by 0xsuu** [comment]
Correct structure will be:
src/Utility/DataGenerator.h
src/Utility/DataGenerator.cpp

Remove src/DataGenerator
Remove src/Utility/main.cpp

**Comment by 0xsuu** [comment]
LGTM

**Comment by 0xsuu** [comment]
@BlueCocoa Please Review

**Comment by 0xsuu** [comment]
@Giolno Next time at someone else to let them review your code.

**Comment by ghost** [comment]
@BlueCocoa @al1enSuu 
plz review this.

File: llama_cpp/src/Utility/DataGenerator.cpp [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+template<>
+void llama::ContinuousData::generateContinuousData<std::string>
+(std::vector<std::string> *targetVector, int count, bool reverse) {
+    std::string generateConinuousString;
+    std::vector<std::string>::iterator it;
+
+    if (reverse) {
+        for (int i = 97; i < count + 97; i++) {
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.cpp:94]: 97 is a, consider A, or all the ascii maybe? @BlueCocoa when generating an ascending order of list of strings, what do you think should be covered?

File: llama_cpp/src/Utility/DataGenerator.h [hunk 10]
```diff
@@ -35,11 +37,28 @@ template<>
...
+};
+
+template<>
+void ContinuousData::generateContinuousData<int>(std::vector<int> *targetVector, int count, bool reverse);
+
+template<>
+void ContinuousData::generateContinuousData<int>(std::vector<int> *targetVector, int count, bool reverse);
+
+template<>
+void ContinuousData::generateContinuousData<int>(std::vector<int> *targetVector, int count, bool reverse);
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.h]: These can be eliminated. Besides, you have three int template specialisations here.

File: llama_cpp/src/Utility/DataGenerator.h [hunk 11]
```diff
@@ -35,11 +37,29 @@ template<>
...
+class ContinuousData {
+ public:
+    ContinuousData() {}
+
+    template<typename T>
+    void generateContinuousData(std::vector<T> *targetVector, int count, bool reverse);
+};
+
+template<>
+void ContinuousData::generateContinuousData<int>(std::vector<int> *targetVector, int count, bool reverse);
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.h]: try delete these

File: llama_cpp/src/Utility/DataGenerator.h [hunk 12]
```diff
@@ -35,11 +37,29 @@ template<>
...
+
+template<>
+void ContinuousData::generateContinuousData<double>(std::vector<double> *targetVector, int count, bool reverse);
+
+template<>
+void ContinuousData::generateContinuousData<std::string>
+(std::vector<std::string> *targetVector, int count, bool reverse);
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.h]: We don't need template specialisation in class definition.

**Comment by 0xsuu** [comment]
LGTM

**Pull request merged**